### PR TITLE
Windows Mixed Reality Gesture Support

### PR DESCRIPTION
**Pull request by StephenHodgson**
Overview
---
- Added Windows Mixed Reality Gesture Support.
- Added windows gesture settings to gesture profile.
image

Changes
---
- Fixes: #2630

### CODE REVIEW

**Comment by SimonDarksideJ** [comment]
Very confused by the changes so far in this PR:
* There's a bunch of changes to the input system, which should be a separate PR IMHO
* A new GestureSettings enum is introduced but not actually used??
* Don't see anything added for gestures?

If we really are expecting to reuse the WMR controller for gestures, I don't see how you are going to be able to configure the input actions for each of those gestures on the controller picture.

**Comment by StephenHodgson** [comment]
I removed the input actions for events like hold, navigation, and manipulation so we didn't have time add anymore actions to the controller, but I'm still not sold on these changes.

One of the main reasons I went this route is because we need to keep track of the controllers and the interaction source id is the same from the gesture recognizer and the interaction manager. Without the interaction manager to raise source detected/lost the gesture controllers would stick around indefinitely.

I did think about pulling out the input system changes but I needed them to happen to make it all work.

In either case it works on HoloLens and WMR no problem. 

I suppose for other gestures on other platforms they'll have their own managers.

**Comment by SimonDarksideJ** [comment]
Agreed, other platforms will have their own gesture controllers, which is another reason it should be it's own type of controller or device.  Really don't like having two independent controllers mashed together like this.

**Comment by StephenHodgson** [comment]
> Really don't like having two independent controllers mashed together like this.

But they're not. They're literally using the same interaction source id.

As for the other platforms having gestures separate from the Controllers, it'll be likely be up to how they have it exposed in their API as to how we will address it.

**Comment by SimonDarksideJ** [comment]
Retracting my comments based off the Shiproom meeting. Full gesture implementation to be reviewed for 2019 build

File: Assets/MixedRealityToolkit/_Core/Definitions/InputSystem/GestureSettings.cs [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+using System;
+
+namespace Microsoft.MixedReality.Toolkit.Core.Definitions.InputSystem
+{
+    /// <summary>
+    /// Copy of Unity's XR.WSA.GestureSettings.
+    /// </summary>
+    [Flags]
+    public enum GestureSettings
```
Comment (davidkline-ms) [inline Assets/MixedRealityToolkit/_Core/Definitions/InputSystem/GestureSettings.cs]: Rename to WindowsMixedRealityGestureSettings? Or do we expect to add additional platform gestures here in the future?

Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Definitions/InputSystem/GestureSettings.cs]: Needs to be cross platform. This isn't platform specific so I won't be renaming it.

Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Definitions/InputSystem/GestureSettings.cs]: I can see us adding to this in the future as more platforms get supported.

**Comment by Ecnassianer** [comment]
> As for the other platforms having gestures separate from the Controllers, it'll be likely be up to how they have it exposed in their API as to how we will address it.

Since we're sourcing gestures from controllers here, we should probably be disambiguating gesture api's back onto our controller classes so the MRTK API is uniform across platforms. The more I see gesture work go in, the more I believe we should create virtual controllers in the case of physical-controllerless hardware like HoloLens or LeapMotion.

**Comment by StephenHodgson** [comment]
@Ecnassianer see #2708

File: Assets/MixedRealityToolkit/_Core/Definitions/InputSystem/MixedRealityInputSystemProfile.cs [hunk 2]
```diff
@@ -63,7 +64,7 @@ public MixedRealityPointerProfile PointerProfile
         /// <summary>
         /// Is the speech Commands Enabled?
         /// </summary>
-        public bool IsSpeechCommandsEnabled => speechCommandsProfile != null && MixedRealityManager.Instance.ActiveProfile.IsInputSystemEnabled && SpeechSystem != null;
+        public bool IsSpeechCommandsEnabled => speechCommandsProfile != null && SpeechSystem != null && MixedRealityManager.Instance.ActiveProfile.IsInputSystemEnabled;
```
Comment (SimonDarksideJ) [inline Assets/MixedRealityToolkit/_Core/Definitions/InputSystem/MixedRealityInputSystemProfile.cs]: This should not be in the Input System. It was removed in the speech PR

Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Definitions/InputSystem/MixedRealityInputSystemProfile.cs]: I don't understand. All I did was flip the order of the check here.

File: Assets/MixedRealityToolkit/_Core/Definitions/InputSystem/GestureSettings.cs [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    {
+        /// <summary>
+        ///   <para>Disable support for gestures.</para>
+        /// </summary>
+        None = 0,
+
+        /// <summary>
+        ///   <para>Enable support for the tap gesture.</para>
+        /// </summary>
+        Tap = 1,
```
[the remark this answers is not in the document]
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Definitions/InputSystem/GestureSettings.cs]: Sure thing!

I just copied the original from the WSA gesture settings but I figured we'd expand this to be our own when new gestures are supported.  I'm not as good at converting these to bit flags, could you help?

[the remark this answers is not in the document]
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Definitions/InputSystem/GestureSettings.cs]: I mainly didn't want to change them bc the file was an exact copy of what Unity's gesture settings enum was. Didn't know if changing the type would affect the transformation.

**Comment by StephenHodgson** [comment]
I should also provide a way to set the gesture settings in the gesture profile.

**Comment by StephenHodgson** [comment]
Okay guys, just need to test this out on the HoloLens then I'll let you know when it's ready for review.

**Comment by davidkline-ms** [comment]
will test Monday am to confirm approval

**Comment by davidkline-ms** [comment]
> It's looking a lot better, but I'm still running into an issue where the pointer is raising two events for some reason.

Is this something that we should delay merging on?

**Comment by StephenHodgson** [comment]
I think the breaking changes getting in are pretty important compared to the duplicate pointer events.

That fix will be easier to merge later.

**Comment by SimonDarksideJ** [comment]
I'ts more critical to get these breaking changes in so we do not have any large impacts later @davidkline-ms else any projects updating later will have larger impacts.
Small things are easier to fix later, plus several other PR's depend on this

**Comment by StephenHodgson** [comment]
@davidkline-ms Okay figured it out. It's not a duplicate event. It's a separate event that's triggered by touch amount being greater than zero. It maps to the WMR controller as Trigger Touched.

**Comment by StephenHodgson** [comment]
Okay just need to fix one little thing with the teleport cursors showing on start and this is good to go

**Comment by StephenHodgson** [comment]
Okay, everything looks good now.

**Comment by StephenHodgson** [comment]
@keveleigh any thoughts on this?

**Comment by StephenHodgson** [comment]
Bippity Boppity Bump

**Comment by davidkline-ms** [comment]
When I test this, I can see Hold Action messages, but when I move my hand, I do not get any Manipulation events (that I can see).

Also, I seem to get Hand Lost a lot...

**Comment by StephenHodgson** [comment]
>Also, I seem to get Hand Lost a lot...

That's because your hand went out of view.

**Comment by StephenHodgson** [comment]
> When I test this, I can see Hold Action messages, but when I move my hand, I do not get any Manipulation events (that I can see).

Strange, I thought I only set it up to use `Manipulation Translate` only.

**Comment by davidkline-ms** [comment]
image

**Comment by davidkline-ms** [comment]
this is what is in the default profile for me (I didn't change it that I can recall)

**Comment by davidkline-ms** [comment]
> That's because your hand went out of view.

Actually, my hand was right in front of me. I barely moved it.

**Comment by StephenHodgson** [comment]
I wonder if the bit shifting is off (haha, it's my weaknesss)

**Comment by StephenHodgson** [comment]
Okay this now correctly handles the events. I had an issue where I was raising the wrong event data

File: Assets/MixedRealityToolkit/_Core/Devices/MixedReality/WindowsMixedRealityController.cs [hunk 4]
```diff
@@ -418,9 +418,24 @@ private void UpdateTriggerData(InteractionSourceState interactionSourceState, Mi
                             if (interactionSourceState.selectPressed)
                             {
                                 InputSystem?.RaiseOnInputDown(InputSource, ControllerHandedness, interactionMapping.MixedRealityInputAction);
+
+                                for (int i = 0; i < InputSource.Pointers.Length; i++)
+                                {
+                                    InputSystem?.RaisePointerDown(InputSource.Pointers[i], ControllerHandedness, interactionMapping.MixedRealityInputAction);
```
[the remark this answers is not in the document]
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Devices/MixedReality/WindowsMixedRealityController.cs]: I'll be sure to update.

[the remark this answers is not in the document]
Comment (keveleigh) [reply Assets/MixedRealityToolkit/_Core/Devices/MixedReality/WindowsMixedRealityController.cs]: I'm not sure how the touch pointer/controller works, but it's probably worth double checking there that this is handled too.

[the remark this answers is not in the document]
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Devices/MixedReality/WindowsMixedRealityController.cs]: I'm pretty sure It's handled correctly.

File: Assets/MixedRealityToolkit/_Core/Devices/MixedReality/WindowsMixedRealityController.cs [hunk 5]
```diff
@@ -418,9 +418,24 @@ private void UpdateTriggerData(InteractionSourceState interactionSourceState, Mi
...
+                                }
                             }
                             else
                             {
+                                for (int i = 0; i < InputSource.Pointers.Length; i++)
+                                {
+                                    InputSystem?.RaisePointerClicked(InputSource.Pointers[i], ControllerHandedness, interactionMapping.MixedRealityInputAction, 1);
```
Comment (keveleigh) [inline Assets/MixedRealityToolkit/_Core/Devices/MixedReality/WindowsMixedRealityController.cs]: The more I think about this change, the less certain I am. It changes the way pointer events work.

At the moment, pointers can be assigned an action to fire their pointer events. This becomes useful for, for example, a sphere-based grab pointer using the grip action. Now, only the trigger can fire those events, as we're explicitly firing them based on the `selectPressed` state.

At the very least, this change is unrelated to gestures, and I'd prefer not to block gestures on it.

**Comment by StephenHodgson** [comment]
Bump

File: Assets/MixedRealityToolkit/_Core/Devices/MixedReality/WindowsMixedRealityController.cs (hunk 5 above)
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Devices/MixedReality/WindowsMixedRealityController.cs]: >**Removed Pointer Up/Down from all scripts raising events outside their respective device manager/controller classes.**

**Comment by StephenHodgson** [comment]
Removed pointer changes and broke out breaking changes into https://github.com/Microsoft/MixedRealityToolkit-Unity/pull/2781 which now blocks this PR

File: Assets/MixedRealityToolkit/_Core/Devices/MixedReality/WindowsMixedRealityDeviceManager.cs [hunk 6]
```diff
@@ -87,19 +353,22 @@ public override void Disable()
         /// Retrieve the source controller from the Active Store, or create a new device and register it
         /// </summary>
         /// <param name="interactionSourceState">Source State provided by the SDK</param>
+        /// <param name="addController">Should the Source be added as a controller if it isn't found?</param>
         /// <returns>New or Existing Controller Input Source</returns>
-        private WindowsMixedRealityController GetOrAddController(InteractionSourceState interactionSourceState)
+        private WindowsMixedRealityController GetController(InteractionSource interactionSourceState, bool addController = true)
```
Comment (keveleigh) [inline Assets/MixedRealityToolkit/_Core/Devices/MixedReality/WindowsMixedRealityDeviceManager.cs]: Should update the parameter name too, to match the new type.

Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Devices/MixedReality/WindowsMixedRealityDeviceManager.cs]: Will do!

**Pull request merged**